certbot/certbot/_internal/storage.py

```diff
@@ -1243,3 +1244,42 @@ def save_successor(self, prior_version: int, new_cert: bytes, new_privkey: bytes
         self.configuration = config_with_defaults(self.configfile)
 
         return target_version
+
+    def truncate(self, current_version: int, previous_version: int,
+                 expiry_cutoff: Optional[datetime.datetime] = None) -> None:
+        """Delete unused historical certificate, chain and key items from the lineage.
+
+        A certificate version will be deleted if it is:
+          1. not the current target, and
+          2. not the previous version (current minus 1), and
+          3. expired (if expiry_cutoff is specified, otherwise all other certs are culled).
+
+        :param expiry_cutoff: The notAfter date after which a certificate is considered expired.
+
+        """
+        # Do not want to delete the current or current-1th target.
+        versions_to_delete = self.available_versions("cert")
+        try:
+            versions_to_delete.remove(current_version)
+            versions_to_delete.remove(previous_version)
+        except ValueError:
+            pass
+        archive = self.archive_dir
+
+        # Optionally, only delete certificates which are already expired.
+        if expiry_cutoff:
+            first_expired_version = next((v for v in versions_to_delete if
+                crypto_util.notAfter(os.path.join(archive, f"cert{v}.pem")) <= expiry_cutoff),
+                None
+            )
+            if first_expired_version:
+                versions_to_delete = [v for v in versions_to_delete if v <= first_expired_version]
+
+        # Delete the remaining lineage items kinds for those certificate versions.
+        for ver in versions_to_delete:
+            logger.debug("Deleting %s/cert%d.pem and related items during clean up",
+                         archive, ver)
+            for kind in ALL_FOUR:
+                item_path = os.path.join(archive, f"{kind}{ver}.pem")
+                if os.path.exists(item_path):
+                    os.unlink(item_path)
```

Comment:
Should these two I/O calls be covered with a broad except and swallowed with a `logger.warning` or something?

Reply:
I'm going to catch `OSError` and cross my fingers that there's not (many) other ways this can go wrong.

Reply:
Just in case GitHub still hasn't made the UI here better, this comment is in response to https://github.com/certbot/certbot/pull/9537#discussion_r1073043043.

I suspect `OSError` is fine. I'm not aware of another exception type that can be raised by the calls here.

If you'd prefer to make it broader though, feel free. I don't think it really hurts anything.